### ISSUE TITLE
custom lint rule for unittest mock.side_effect

### ISSUE THREAD

**Issue by upstartjohnvandivier** [astral-sh/ruff#6199]
as a developer, I would like ruff to lint two improper uses of .side_effect, so that I can avoid hard-to-troubleshoot issues in writing unit tests and I can improve readility and test performance:
1. side_effect arrays with a length less than two
2. side_effect arrays where each element is the same value

eg:
```python
# error
foo.side_effect = [bar]
foo.side_effect = [bar, bar]

# prefer to above
foo.return_value = bar

# okay usage
foo.side_effect = [bar, baz]
```

**Comment by charliermarsh** [comment]
Can you refresh my memory, what's the problem with these usages?

```python
# error
foo.side_effect = [bar]
foo.side_effect = [bar, bar]
```

**Comment by upstartjohnvandivier** [comment]
bar can be mutated byref in this implementation

for example, if bar is a unittest.Mock and it's called once in a given spec, and we assert called once, it may fail if it was also called once in another spec.

in addition, this usage is intent breaking; if you are setting it to return a value that isn't a side effect. that's the use case for `. return_value`

**Comment by upstartjohnvandivier** [comment]
I just added a feature request to pytest-mock to similar effect: It would outright fail a spec on improper usage rather than catching it at linting time. I would eagerly point out two benefits of adding this to ruff invariant to whether they accept the feature request:
1. The actual projects where I use ruff and pytest don't currently include pytest-mock. I would have to add an additional dep if ruff doesn't support this.
2. Ruff is many times faster than running the unit test suite. So, for rapid developer feedback it would be nice to have this rule whether or not the test will later fail.

**Comment by upstartjohnvandivier** [comment]
Python language proposal created at https://discuss.python.org/t/feat-unittest-mock-side-effect-fails-on-improper-usage/31003

Support is welcome

**Comment by upstartjohnvandivier** [comment]
> The meaning changes when you use return_value = x instead of side_effect = [x]
Right, and in the codebases I've worked in, the StopIteration error has never been intended

> Additionally, side-effects have different meanings for exception types, e.g.
Yes. This request is not to mark those cases as improper. Notice that those cases are proper side effects and they are not byref constant references

<img width="687" alt="image" src="https://github.com/astral-sh/ruff/assets/102185031/83e12319-327d-4e45-ab5c-555c2aa27411">

**Comment by upstartjohnvandivier** [comment]
> The meaning changes when you use return_value

Yes, the StopIteration error is one of the hard to troubleshoot test issues that I referred to. I've never seen it actually intended. It generally represents a mis-implemented test in the codebases that I work in, rather than an expected effect of the method under test.

> Additionally, side-effects have different meanings for exception
Yes. This is the main and proper use case for .side_effect, not a special additional consideration. This feature request does not mark  request is not to mark those cases as improper. Notice that those cases are proper side effects

z = bar() doesn't `ValueError` it throws as a side effect. z isn't assigned a value of `ValueError`

Further, `bar.side_effect = [ValueError(), ValueError(), ValueError()]` would be extremely weird. I've never seen a need for that. tests will generally error on the first invocation

I think `bar.side_effect = ValueError()` is generally preferred to `bar.side_effect = [ValueError()]`, so I think the length check lgtm, but if we want to apply a special exception for a list of errors I would still consider that better than the state without any lint rule. Did that make sense?
